src/java.base/share/classes/java/util/concurrent/ConcurrentSkipListSet.java

```diff
@@ -176,6 +178,8 @@ public ConcurrentSkipListSet<E> clone() {
             ConcurrentSkipListSet<E> clone =
                 (ConcurrentSkipListSet<E>) super.clone();
             clone.setMap(new ConcurrentSkipListMap<E,Object>(m));
+            // Needed to ensure safe publication of setMap()
+            VarHandle.releaseFence();
```

Comment:
@DougLea This might be interesting to you. 👍

Reply:
I wonder if we could just change it to return `new ConcurrentSkipListSet<>(m)`. COWAL could be changed to return a new object too.

Reply:
The set only has a map field, and AbstractSet does not define any additional field. The map should be fine too - two fields in AbstractMap are cleared when cloning happens, so recreating a map from a constructor should have the same effect. (Note a significant field `head` is not cleared upon clone, but seems immediately replaced later in `buildFromSorted`). Both should still be fine with this new return values.

Reply:
As others have mentioned, clone() needs to be re-checked wrt issuing final field fences. But can be conservatively done so here anyway.

Reply:
@DougLea Yeah. I'd be surprised if `clone()`-invocations were specially tracked to get a trailing release fence inserted before handing out the instance.